packaging: add native debian package

cjp256 opened the pull request:
For development purposes, add native debian packaging.

- Add build-deb.sh to /scripts/

- Add debian package metadata to /packaging/debian

- Add a starter manpage to /doc in markdown format, using pandoc to convert to manpage.

- Add github workflow to validate debs.

- Move azure-nvme-id.spec to /packaging/azurelinux.  It may work for other RPM distros too, but will revisit structure upon validating Fedora, etc.

nmeyerhans commented on CMakeLists.txt:
```diff
@@ -41,6 +38,39 @@ configure_file(
+set(UDEV_RULES_INSTALL_DIR "/lib/udev/rules.d")
+
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/doc/azure-nvme-id.1
+        DESTINATION ${MANPAGES_INSTALL_DIR}/man1)
```
nit: IMO, this change should be in a commit by itself, since it's not specific to the .deb packaging at all. Just for the sake of cleanliness/organization.

cjp256 replied:
agreed, I actually already did that here https://github.com/Azure/azure-nvme-utils/pull/17  :D  need to rebase this PR

nmeyerhans commented on .github/workflows/debs.yml:
```diff
@@ -0,0 +1,44 @@
+    - name: Build debs
+      run: |
+        ./scripts/build-deb.sh
+    - name: Install debs
```
It'd be good to invoke `lintian` in a GitHub workflow step. We want to ensure we're compliant with distro policies and best practice.  It should be sufficient for now to install the `lintian` package and then invoke `lintian out/*.changes`

cjp256 replied:
added lintian check, but in my local build I'm seeing a warning during build I don't see when running lintian separately:

```
Now running lintian azure-nvme-utils_0.1.2+git393c9ca_amd64.changes ...
W: azure-nvme-utils-dbgsym: elf-error In program headers: Unable to find program interpreter name [usr/lib/debug/.build-id/b1/e816f72bb315c47eb8f759854aca258476017a.debug]
Finished running lintian.
```

Any idea?

nmeyerhans replied:
I'm not seeing this error in current builds in a sid environment, though I was when there were leftover files in packaging/debian/... from a previous build done when that was symlinked to the top-level debian subdirectory.  So I think this particular lintian alert is a non-issue.  However, there do seem to be some new ones...

cjp256 replied:
rather, seems sorted when building on debian sid

nmeyerhans commented on scripts/build-deb.sh:
```diff
@@ -0,0 +1,17 @@
+debuild
+
+mkdir -p out
+mv ../azure-nvme-utils_99.99.99* out/
```
This glob misses `azure-nvme-utils-dbgsym_99.99.99*` packages.

cjp256 replied:
fixed, thanks!

nmeyerhans commented on packaging/debian/changelog:
```diff
@@ -0,0 +1,5 @@
+azure-nvme-utils (99.99.99) UNRELEASED; urgency=low
```
I'd use a version like 0.0~1, to ensure that it's always sorted as an earlier version than what may be available otherwise.  We don't want people accidentally upgrading to dev versions (it can still be installed if people really mean it)

cjp256 replied:
i switched the specs to using a dynamic version determined from the git tree and was thinking about doing the same here instead of a hardcoded fake version

cjp256 replied:
updated to determine it dynamically, e.g. azure-nvme-utils_0.1.2+git393c9ca_amd64.deb

nmeyerhans reviewed (changes requested):
Looks good over all, just a few relatively minor suggestions.

cjp256 committed:
packaging: add native debian package

For development purposes, add native debian packaging.

- Add debian package metadata to /packaging/debian

- Add build-deb.sh to /scripts/ to build the debs

- Add github workflow to validate debs

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

nmeyerhans commented on packaging/debian/control:
```diff
@@ -0,0 +1,18 @@
+Depends: ${misc:Depends}, ${shlibs:Depends}
+Description: auto-generated package by debmake
+ This Debian binary package was auto-generated by the
+ debmake(1) command provided by the debmake package.
```
The description should get filled out for packages that will actually be distributed, and will definitely need to be completed for submission to Debian.

For dev builds as done here, it won't matter, so this isn't a blocker for this PR IMO.

The structure of the description is described in the [debian policy](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-description)

cjp256 replied:
fixed, thanks!

nmeyerhans commented on scripts/build-deb.sh:
```diff
@@ -0,0 +1,30 @@
+    git_email="$(git config user.email)"
+    export DEBEMAIL="${git_user} <${git_email}>"
+fi
+dch --create -v "${deb_version}" --package "azure-nvme-utils" "development build: ${git_version}"
```
`dch` looks at the Maintainer and Uploader fields in debian/control to determine whether a given change is being made by a maintainer or not. (see [the docs](https://manpages.debian.org/bookworm/devscripts/dch.1.en.html#nmu))  When I run this script, the changelog is generated with my email address and it's recognized as a non-maintainer-upload (NMU) because my address isn't listed in `debian/control`.  So the resulting packages generate a lintian warning:
```plain
W: azure-nvme-utils source: no-nmu-in-changelog [debian/changelog:1]
W: azure-nvme-utils source: source-nmu-has-incorrect-version-number 0.1.3+gitd60d0e8DIRTY [debian/changelog:1]
```

For the dev builds, we should suppress these warnings by adding the following to `packaging/debian/source/lintian-overrides`:

```
source: no-nmu-in-changelog
source: source-nmu-has-incorrect-version-number
```

cjp256 replied:
done, thanks!

nmeyerhans commented on .gitignore:
```diff
@@ -73,6 +73,9 @@ _deps
 build/
-udev/80-azure-nvme.rules
-src/version.h
+debian/
```
This should be `/debian`, or else you'll end up ignoring `packaging/debian` as well, which is not what you want.

The line above it should probably also be `/build`.

cjp256 replied:
fixed, thanks!

nmeyerhans reviewed (changes requested):
lintian detects the following warnings from groff on sid:

```plain
W: azure-nvme-utils: groff-message troff:<standard input>:50: warning: cannot select font 'C' [usr/share/man/man1/azure-nvme-id.1.gz:2]
W: azure-nvme-utils: groff-message troff:<standard input>:5: warning: cannot select font 'CB' [usr/share/man/man1/azure-nvme-id.1.gz:1]
W: azure-nvme-utils: groff-message troff:<standard input>:60: warning: cannot select font 'C' [usr/share/man/man1/azure-nvme-id.1.gz:3]
W: azure-nvme-utils: groff-message troff:<standard input>:72: warning: cannot select font 'C' [usr/share/man/man1/azure-nvme-id.1.gz:4]
```

This appears to be a [pandoc issue](https://github.com/jgm/pandoc/issues/9020) and may be fixed upstream.  I'm not sure if this can be worked around in the input file or not, but it's probably worth a try.  You can see the warnings yourself with

```plain
$ man -l obj-x86_64-linux-gnu/doc/azure-nvme-id.1  > /dev/null
troff:<standard input>:5: warning: cannot select font 'CB'
troff:<standard input>:50: warning: cannot select font 'C'
troff:<standard input>:60: warning: cannot select font 'C'
troff:<standard input>:72: warning: cannot select font 'C'
```

cjp256 commented:
> lintian detects the following warnings from groff on sid:
> 
> ```
> W: azure-nvme-utils: groff-message troff:<standard input>:50: warning: cannot select font 'C' [usr/share/man/man1/azure-nvme-id.1.gz:2]
> W: azure-nvme-utils: groff-message troff:<standard input>:5: warning: cannot select font 'CB' [usr/share/man/man1/azure-nvme-id.1.gz:1]
> W: azure-nvme-utils: groff-message troff:<standard input>:60: warning: cannot select font 'C' [usr/share/man/man1/azure-nvme-id.1.gz:3]
> W: azure-nvme-utils: groff-message troff:<standard input>:72: warning: cannot select font 'C' [usr/share/man/man1/azure-nvme-id.1.gz:4]
> ```
> 
> This appears to be a [pandoc issue](https://github.com/jgm/pandoc/issues/9020) and may be fixed upstream. I'm not sure if this can be worked around in the input file or not, but it's probably worth a try. You can see the warnings yourself with
> 
> ```
> $ man -l obj-x86_64-linux-gnu/doc/azure-nvme-id.1  > /dev/null
> troff:<standard input>:5: warning: cannot select font 'CB'
> troff:<standard input>:50: warning: cannot select font 'C'
> troff:<standard input>:60: warning: cannot select font 'C'
> troff:<standard input>:72: warning: cannot select font 'C'
> ```

Addressing with https://github.com/Azure/azure-nvme-utils/pull/27

nmeyerhans commented on packaging/debian/control:
```diff
@@ -0,0 +1,18 @@
+Homepage: https://github.com/Azure/azure-nvme-utils
+Rules-Requires-Root: no
+#Vcs-Git: https://salsa.debian.org/debian/azure-nvme.git
+#Vcs-Browser: https://salsa.debian.org/debian/azure-nvme
```
nit: either delete these or uncomment them and update them to point to the github repos

cjp256 replied:
done, thanks!

nmeyerhans reviewed (approved):
Any reason not to squash cc097bcc0c9ffdeb5986dc24800c989c37ed13e0 ("fix description in debian/control") into 49e60b6440b00400909165cce777dca16a06c5d1 ("packaging: add native debian package")?

cjp256 commented:
> Any reason not to squash [cc097bc](https://github.com/Azure/azure-nvme-utils/commit/cc097bcc0c9ffdeb5986dc24800c989c37ed13e0) ("fix description in debian/control") into [49e60b6](https://github.com/Azure/azure-nvme-utils/commit/49e60b6440b00400909165cce777dca16a06c5d1) ("packaging: add native debian package")?

Squashed. Sometimes I try to be github-review-friendly and squash on merge. :)
